zeus21/sfrd.py

```diff
@@ -57,258 +106,429 @@ def __init__(self, Cosmo_Parameters, ClassCosmo, Astro_Parameters, HMF_interpola
+        ### STEP 4: LW correction to Pop III gammas
+        if Astro_Parameters.USE_POPIII == True:
+            if Astro_Parameters.USE_LW_FEEDBACK == True:
+                #get the zero-lag correlation function (zero distance separation)
+                xi_RR_CF_zerolag = np.copy(ClassCosmo.pars['xi_RR_CF'][:,:,0])
+
+                #compute LW coefficients for Pop II and III stars
+                coeff1LWzp_II, coeff2LWzpRR_II = J_LW_Discrete(Astro_Parameters, Cosmo_Parameters, ClassCosmo, self.zintegral, 2, self.Rtabsmoo, SFRD_II_interp, SFRD_III_cnvg_interp)
+                coeff1LWzp_III, coeff2LWzpRR_III = J_LW_Discrete(Astro_Parameters, Cosmo_Parameters, ClassCosmo, self.zintegral, 3, self.Rtabsmoo, SFRD_II_interp, SFRD_III_cnvg_interp)
+
+                # Corrections WITH Rmax smoothing
+                deltaGamma_R = 1 / np.transpose([SFRD_III_cnvg_interp(self.zintegral)])
+                deltaGamma_R *= np.array([dSFRDIII_dJ(Astro_Parameters, Cosmo_Parameters, ClassCosmo, HMF_interpolator, J21LW_interp, np.array([self.zintegral]).T, np.array([self.zintegral]).T, ClassCosmo.pars['v_avg'])]).T
+                deltaGamma_R = deltaGamma_R * (coeff1LWzp_II * coeff2LWzpRR_II * self.gamma_II_index2D + coeff1LWzp_III * coeff2LWzpRR_III * self.gamma_III_index2D) * 1e21
+
+                #choose only max of r and R; since growth factors cancel out, none are used here
+                xi_R_maxrR = np.tril(np.ones_like(xi_RR_CF_zerolag)) * np.transpose([np.diag(xi_RR_CF_zerolag)])
+                xi_R_maxrR  = xi_R_maxrR  + np.triu(xi_RR_CF_zerolag, k = 1)
+
+                self.deltaGamma_R_Matrix = xi_R_maxrR.reshape(len(self.Rtabsmoo), 1, len(self.Rtabsmoo)) * (deltaGamma_R * self.dlogRR * self.Rtabsmoo).reshape(1, len(self.zintegral), len(self.Rtabsmoo))
+                deltaGamma_R_z = np.transpose(   np.sum(self.deltaGamma_R_Matrix, axis = 2) / np.transpose([np.diagonal(xi_RR_CF_zerolag[:,:])])    )
+                deltaGamma_R_z[ self.gamma_III_index2D == 0 ] = 0 #don't correct gammas if gammas are zero
+                self.deltaGamma_R_z = deltaGamma_R_z
+                self.gamma_III_index2D += deltaGamma_R_z #correct Pop III gammas with LW correction factor
+                
+
+        #####################################################################################################
+        ### STEP 5: Lyman-Alpha Anisotropies
+        
+        #Makes heavy use of broadcasting to make computations faster
+        #3D cube will be summed over one axis. Dimensions are (z,R,n) = (64, 45, 21)
 
-                nu0 = Cosmo_Parameters.delta_crit_ST/sigmacurr #the EPS delta = 0 result. Note 1/sigmacurr and not sigmamod
-                nu0[indextoobig]=1.0 #set to 1 to avoid under/overflows, we don't sum over those masses since they're too big
+        self.coeff1LyAzp = (1+self.zintegral)**2/(4*np.pi)
 
+        nuLYA = np.geomspace(constants.freqLyA, constants.freqLyCont, 128)
```

Comment:
There's a difference in computation time here; because SED_LyA computes things elementwise for each nu, feeding SED_LyA a cube of frequencies takes slightly longer than interpolating it and feeding the cube of frequencies into the interpolator.